From aa46e7502b3a4745a3c97ea83902d053e6a45be1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ond=C5=99ej=20Jirman?= <megi@xff.cz>
Date: Fri, 28 Jan 2022 16:57:31 +0100
Subject: [PATCH 368/456] power: supply: rk818: Don't ever set input current
 limit to 0

All USB power sources will allow the at least 500 mA to be consumed.
Without this patch, when fusb302 updates its psy after boot, without
properly runnning the usb2 phy based charger type detection, at least
we don't drop the input current limit to 0, preventing charging
completely.

Signed-off-by: Ondrej Jirman <megi@xff.cz>
---
 drivers/power/supply/power_supply_core.c | 10 +++++++++-
 drivers/power/supply/rk818_charger.c     |  2 +-
 include/linux/power_supply.h             |  2 ++
 3 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/power/supply/power_supply_core.c b/drivers/power/supply/power_supply_core.c
index 2abd4d3810db..91aa09e2c213 100644
--- a/drivers/power/supply/power_supply_core.c
+++ b/drivers/power/supply/power_supply_core.c
@@ -376,7 +376,7 @@ int power_supply_is_system_supplied(void)
 }
 EXPORT_SYMBOL_GPL(power_supply_is_system_supplied);
 
-int power_supply_set_input_current_limit_from_supplier(struct power_supply *psy)
+int power_supply_set_input_current_limit_from_supplier_min(struct power_supply *psy, int minimum)
 {
 	union power_supply_propval val = {0,};
 	struct class_dev_iter iter;
@@ -405,9 +405,17 @@ int power_supply_set_input_current_limit_from_supplier(struct power_supply *psy)
 	if (ret)
 		return ret;
 
+	val.intval = max(val.intval, minimum);
+
 	return psy->desc->set_property(psy,
 				POWER_SUPPLY_PROP_INPUT_CURRENT_LIMIT, &val);
 }
+EXPORT_SYMBOL_GPL(power_supply_set_input_current_limit_from_supplier_min);
+
+int power_supply_set_input_current_limit_from_supplier(struct power_supply *psy)
+{
+	return power_supply_set_input_current_limit_from_supplier_min(psy, 0);
+}
 EXPORT_SYMBOL_GPL(power_supply_set_input_current_limit_from_supplier);
 
 int power_supply_set_battery_charged(struct power_supply *psy)
diff --git a/drivers/power/supply/rk818_charger.c b/drivers/power/supply/rk818_charger.c
index 097f30b7689d..d4ed7b1d558a 100644
--- a/drivers/power/supply/rk818_charger.c
+++ b/drivers/power/supply/rk818_charger.c
@@ -234,7 +234,7 @@ static void rk818_usb_power_external_power_changed(struct power_supply *psy)
 {
         struct rk818_charger *cg = power_supply_get_drvdata(psy);
 
-	power_supply_set_input_current_limit_from_supplier(cg->usb_psy);
+	power_supply_set_input_current_limit_from_supplier_min(cg->usb_psy, 500000);
 }
 
 static enum power_supply_property rk818_usb_power_props[] = {
diff --git a/include/linux/power_supply.h b/include/linux/power_supply.h
index e218041cc000..d0670c62320b 100644
--- a/include/linux/power_supply.h
+++ b/include/linux/power_supply.h
@@ -599,6 +599,8 @@ extern void power_supply_changed(struct power_supply *psy);
 extern int power_supply_am_i_supplied(struct power_supply *psy);
 extern int power_supply_set_input_current_limit_from_supplier(
 					 struct power_supply *psy);
+extern int power_supply_set_input_current_limit_from_supplier_min(
+					 struct power_supply *psy, int minimum);
 extern int power_supply_set_battery_charged(struct power_supply *psy);
 
 #ifdef CONFIG_POWER_SUPPLY
-- 
2.34.1

